### PR TITLE
Problem: mempool don't respect gas wanted returned by ante handler

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- support custom gas wanted in mempool

# Description

Closes: #XXXX

---

## Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] provided a link to the relevant issue or specification
* [ ] reviewed "Files changed" and left comments if necessary
* [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
* [ ] added a changelog entry to `CHANGELOG.md`
* [ ] updated the relevant documentation or specification, including comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] confirmed all CI checks have passed

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic, API design and naming, documentation is accurate, tests and test coverage

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -60,6 +60,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 * (baseapp) [#20144](https://github.com/cosmos/cosmos-sdk/pull/20144) Remove txs from mempool when antehandler fails in recheck.
 * (baseapp) [#20107](https://github.com/cosmos/cosmos-sdk/pull/20107) Avoid header height overwrite block height.
 * (cli) [#20020](https://github.com/cosmos/cosmos-sdk/pull/20020) Make bootstrap-state command support both new and legacy genesis format.
+* [#]() mempool respect gas wanted returned by ante handler
```
Comment (yihuang) [inline CHANGELOG.md]:
```suggestion
* [#507](https://github.com/crypto-org-chain/cosmos-sdk/pull/507) mempool respect gas wanted returned by ante handler
```

File: types/mempool/priority_nonce.go [hunk 2]
```diff
@@ -382,7 +395,7 @@ func (mp *PriorityNonceMempool[C]) reorderPriorityTies() {
 		if mp.priorityCounts[key.priority] > 1 {
 			newKey := key
 			newKey.weight = senderWeight(mp.cfg.TxPriority, key.senderElement)
-			reordering = append(reordering, reorderKey[C]{deleteKey: key, insertKey: newKey, tx: node.Value.(sdk.Tx)})
+			reordering = append(reordering, reorderKey[C]{deleteKey: key, insertKey: newKey, tx: node.Value.(MempoolTx).Tx})
```
Comment (mmsqe) [inline types/mempool/priority_nonce.go]: seems it's not a mempool.MempoolTx

Comment (yihuang) [reply types/mempool/priority_nonce.go]: yes, priority index set `sdk.Tx`, sender index set `MempoolTx`

**Review by mmsqe — COMMENTED** [review]
seems also need align [ProcessProposal](https://github.com/crypto-org-chain/cosmos-sdk/blob/release/v0.50.x/baseapp/abci_utils.go#L397) totalTxGas check

**Comment by yihuang** [comment]
> seems also need align [ProcessProposal](https://github.com/crypto-org-chain/cosmos-sdk/blob/release/v0.50.x/baseapp/abci_utils.go#L397) totalTxGas check

good catch, fixed

**Review by mmsqe — APPROVED** [review]

File: types/mempool/mempool.go [hunk 3]
```diff
@@ -7,11 +7,30 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
+type MempoolTx struct {
```
Comment (mmsqe) [inline types/mempool/mempool.go]: not sure if we need expose this

Comment (yihuang) [reply types/mempool/mempool.go]: it's part of mempool interface

**Pull request merged**